source/common/router/router.cc

```diff
@@ -547,8 +549,14 @@ Http::FilterHeadersStatus Filter::decodeHeaders(Http::HeaderMap& headers, bool e
   retry_state_ =
       createRetryState(route_entry_->retryPolicy(), headers, *cluster_, config_.runtime_,
                        config_.random_, callbacks_->dispatcher(), route_entry_->priority());
-  do_shadowing_ = FilterUtility::shouldShadow(route_entry_->shadowPolicy(), config_.runtime_,
-                                              callbacks_->streamId());
+
+  // Determine which shadow policies to use. It's possible that we don't do any shadowing due to
+  // runtime keys.
+  for (const auto& shadow_policy : route_entry_->shadowPolicies()) {
+    if (FilterUtility::shouldShadow(shadow_policy, config_.runtime_, callbacks_->streamId())) {
+      active_shadow_policies_.push_back(std::cref<ShadowPolicyPtr>(shadow_policy));
```

Comment:
i'd store a ref to the underlying policy here as well: ownership semantics should be the same, but its more explicit that we're never storing a nullptr

Reply:
I'm not quite following this comment, what do you mean by store a ref to the underlying policy? I used `std::cref`/`std::reference_wrapper` here since you can't have a vector of references but wanted to avoid duplicating/recreating the pointers. `std::reference_wrapper` can't hold a `nullptr`: https://godbolt.org/z/zUPTEe

Reply:
Your example doesn't work because you cant bind std::cref to an rvalue reference: https://en.cppreference.com/w/cpp/utility/functional/ref (note how the move ctor is deleted). If you bind it to a temporary first it works: https://godbolt.org/z/fEB456

Basically my suggestion was to do `std::cref<ShadowPolicy>` and do `active_shadow_policies_.push_back(std::cref(shadow_policy.get())`. 

All that said it's a very minor thing, so I don't feel too strongly about it.

Reply:
Ah gotcha, I thought your comment was about switching from `std::cref` to another paradigm. Sure can update